### ISSUE TITLE
Add CuDNN deterministic mode

### ISSUE THREAD

**Issue by AlexDBlack** [eclipse/deeplearning4j#5089]
CuDNN implements a number of different algorithms for CNN etc, and (based on our use of CuDNN in DL4J) automatically selects which algorithm to use.
However, many of those algorithms are non-deterministic (but may be faster than the deterministic implementations) - this non-determinism can make it challenging to debug issues that require reproducibility between runs.

Here's the scores of the same net run 5 times on the same data with cudnn:
![image](https://user-images.githubusercontent.com/2360237/39738781-7a8e3cc6-52d0-11e8-9064-18fdfcaa0075.png)

Config from here (minus dropout): https://github.com/deeplearning4j/deeplearning4j/issues/5068

See also, for example:
https://github.com/pytorch/pytorch/issues/114
https://github.com/tensorflow/tensorflow/issues/12871

**Comment by agibsonccc** [comment]
Closing — the CUDA backend has been completely rewritten with new memory management, DSP execution engine, CUDA graph replay, and Triton compilation. If you still experience GPU-related issues with the upcoming release, please open a new issue.
